### PR TITLE
2.38-DHIS2-12326 Event analytics always returns metadata for all options in option set

### PR DESCRIPTION
**Pull request by d-bernat**
Current behaviour

When a data element with an option set is requested on /analytics/events/query metadata is returned for all options in the option set. Even when only a few options are specifically requested, e.g. dimension=K6uUAvq500H:IN:A00;A000, metadata for all options are returned.

Example where a single option has been specified. Still metadata for all 14000+ options are returned:
https://play.dhis2.org/dev/api/29/analytics/events/query/eBAyeGv0exc.json?dimension=pe:LAST_3_MONTHS&dimension=ou:ImspTQPwCqd&dimension=Zj7UnCAulEk.K6uUAvq500H:IN:A00&displayProperty=SHORTNAME&outputType=EVENT&desc=eventdate&pageSize=100&page=1

Example request where no options are specified and page size is set to 100. Still metadata for all 14000+ options are returned:
https://play.dhis2.org/dev/api/analytics/events/query/eBAyeGv0exc.json?dimension=pe:LAST_3_MONTHS&dimension=ou:ImspTQPwCqd&dimension=Zj7UnCAulEk.K6uUAvq500H&outputType=EVENT&desc=eventdate&pageSize=100&page=1

Expected behaviour

When e.g. two options are requested, only return metadata for those two. If the dimension is requested without specified options then only return metadata for the options that exist in the (paged) data response.

### CODE REVIEW

**Review by gnespolino — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractAnalyticsService.java [hunk 1]
```diff
@@ -264,33 +310,62 @@ protected void addMetadata( EventQueryParams params, Grid grid )
...
+
+        if ( !params.isSkipData() )
+        {
+            itemOptions.forEach( option -> metadataItemMap.put( option.getUid(),
+                new MetadataItem( option.getDisplayName(), includeDetails ? option.getUid() : null,
+                    option.getCode() ) ) );
+        }
+        else
+        {
+            params.getItemOptions().stream()
```
Comment (maikelarabori) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractAnalyticsService.java:348]: Could you add a comment here just to give a quick overview of what's going on in this filtering and the reason for it?

**Review by maikelarabori — APPROVED** [review]
Hi @d-bernat, looks good. Just added some minor comments. Please have a look at them before merging.

Thanks,
Maikel

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/JdbcAnalyticsManager.java [hunk 2]
```diff
@@ -600,7 +600,7 @@ private String getFirstOrLastValueSubquerySql( DataQueryParams params, Date earl
     {
         Period period = params.getLatestPeriod();
 
-        List<String> cols = Lists.newArrayList( "year", "pestartdate", "peenddate", "level", "daysxvalue", "daysno",
+        List<String> cols = Lists.newArrayList( "year", "pestartdate", "peenddate", "oulevel", "daysxvalue", "daysno",
```
[the remark this answers is not in the document]
Comment (d-bernat) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/JdbcAnalyticsManager.java:603]: yes, I it is.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractAnalyticsService.java (hunk 1 above)
Comment (d-bernat) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractAnalyticsService.java:348]: done

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractAnalyticsService.java [hunk 3]
```diff
@@ -243,13 +252,50 @@ protected void addMetadata( EventQueryParams params, Grid grid )
...
+        for ( int i = 0; i < grid.getHeaders().size(); ++i )
+        {
+            GridHeader gridHeader = grid.getHeaders().get( i );
+
+            if ( gridHeader.hasOptionSet() )
+            {
+                final int columnIndex = i;
+
+                options.addAll( gridHeader
+                    .getOptionSetObject()
```
[the remark this answers is not in the document]
Comment (d-bernat) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractAnalyticsService.java:274]: the test is there, see line 269 if ( gridHeader.hasOptionSet() )

**Pull request merged**